File: dsa/arrays,hashing/128. Longest Consecutive Sequence.py
"""
LC 128 — Longest Consecutive Sequence
Pattern: Hash Set + Sequence Start Detection
Difficulty: Medium
Time: __ min
Approach: Insert all numbers into a set for O(1) lookup.
          A number starts a sequence only if (num - 1) is not present.
          From each valid start, keep extending forward: num + 1, num + 2...
          Track the maximum sequence length.
Time Complexity: O(n)
Space Complexity: O(n)
"""
from typing import List

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        st = set(nums)
        longest = 0

        for num in st:
            # only start counting if num is the beginning of a sequence
            if num - 1 not in st:
                curr_num = num
                curr_len = 1

                while curr_num + 1 in st:
                    curr_num += 1
                    curr_len += 1

                longest = max(longest, curr_len)

        return longest

File: dsa/arrays,hashing/128. Longest Consecutive Sequence.cpp
/*
LC 128 — Longest Consecutive Sequence
Pattern: Hash Set + Sequence Start Detection
Difficulty: Medium
Time: __ min
Approach: Insert all numbers into an unordered_set for O(1) lookup.
          A number starts a sequence only if (num - 1) is not present.
          From each valid start, keep extending forward: num + 1, num + 2...
          Track the maximum sequence length.
Time Complexity: O(n)
Space Complexity: O(n)
*/

#include<bits/stdc++.h>
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> st(nums.begin(), nums.end());
        int longest = 0;

        for (int num : st) {
            // only start counting if num is the beginning of a sequence
            if (st.find(num - 1) == st.end()) {
                int currNum = num;
                int currLen = 1;

                while (st.find(currNum + 1) != st.end()) {
                    currNum++;
                    currLen++;
                }

                longest = max(longest, currLen);
            }
        }

        return longest;
    }
};